zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -709,33 +710,93 @@ where
+                InputsErrT,
+                FeeRuleT::Error,
+            >,
+        > {
             let address_metadata = known_addrs
-                .get(utxo.recipient_address())
-                .ok_or_else(|| Error::AddressNotRecognized(*utxo.recipient_address()))?
+                .get(addr)
+                .ok_or(Error::AddressNotRecognized(*addr))?
                 .clone()
-                .ok_or_else(|| Error::NoSpendingKey(utxo.recipient_address().encode(params)))?;
+                .ok_or_else(|| Error::NoSpendingKey(addr.encode(params)))?;
 
             let secret_key = usk
                 .transparent()
                 .derive_secret_key(address_metadata.scope(), address_metadata.address_index())
                 .unwrap();
 
-            builder.add_transparent_input(
-                secret_key,
+            utxos_spent.push(outpoint.clone());
+            builder.add_transparent_input(secret_key, outpoint, utxo)?;
+
+            Ok(())
+        };
+
+        for utxo in proposal_step.transparent_inputs() {
+            add_transparent_input(
+                utxo.recipient_address(),
                 utxo.outpoint().clone(),
                 utxo.txout().clone(),
             )?;
         }
-        utxos
+        for input_ref in proposal_step.prior_step_inputs() {
+            match input_ref.output_index() {
+                proposal::StepOutputIndex::Payment(i) => {
+                    // We know based upon the earlier check that this must be a transparent input,
+                    // We also know that transparent outputs for that previous step were added to
+                    // the transaction in payment index order, so we can use dead reckoning to
+                    // figure out which output it ended up being.
+                    let (prior_step, result) = &prior_step_results[input_ref.step_index()];
+                    let recipient_address = match &prior_step
+                        .transaction_request()
+                        .payments()
+                        .get(&i)
+                        .expect("Payment step references are checked at construction")
+                        .recipient_address
+                    {
+                        Address::Transparent(t) => Some(t),
+                        Address::Unified(uaddr) => uaddr.transparent(),
+                        _ => None,
+                    }
+                    .ok_or(Error::ProposalNotSupported)?;
+                    let outpoint = OutPoint::new(
+                        result.transaction().txid().into(),
+                        u32::try_from(
+                            prior_step
+                                .payment_pools()
+                                .iter()
+                                .filter(|(_, pool)| pool == &&PoolType::Transparent)
+                                .take_while(|(j, _)| j <= &&i)
+                                .count()
+                                - 1,
```

Comment:
Hmm, I'm not a huge fan of this dead reckoning, given how we randomize shielded output orders and could in theory also randomize transparent output orders, but it's fine for now.